hierarchicalforecast/utils.py

```diff
@@ -557,3 +565,202 @@ def samples_to_quantiles_df(samples: np.ndarray,
+    
+    # We need the empirical covariance, the off-diagonal sum of the variance of 
+    # the empirical correlation matrix and the off-diagonal sum of the squared 
+    # empirical correlation matrix.
+    W = np.zeros((n_timeseries, n_timeseries), dtype=np.float64).T
+    sum_var_emp_corr = np.float64(0.0)
+    sum_sq_emp_corr = np.float64(0.0)
+    epsilon = np.float64(2e-8)
+    for i in prange(n_timeseries):
+        not_nan_mask_i = not_nan_mask[i]
+        for j in range(i + 1):
+            not_nan_mask_j = not_nan_mask[j]
+            not_nan_mask_ij = not_nan_mask_i & not_nan_mask_j   
+            n_samples = np.sum(not_nan_mask_ij)
+            # Only compute if we have enough non-nan samples in the time series pair
+            if n_samples > 1:
+                # Masked residuals
+                residuals_i = residuals[i][not_nan_mask_ij]
+                residuals_j = residuals[j][not_nan_mask_ij]
+                residuals_i_mean = np.mean(residuals_i)
+                residuals_j_mean = np.mean(residuals_j)
+                X_i = (residuals_i - residuals_i_mean)
+                X_j = (residuals_j - residuals_j_mean)
+                # Empirical covariance
+                factor_emp_cov = np.float64(1 / (n_samples - 1))
+                W[i, j] = factor_emp_cov * np.sum(X_i * X_j)
+                # Off-diagonal sums
+                if i != j:
+                    factor_var_emp_cor = np.float64(n_samples / (n_samples - 1)**3)
+                    residuals_i_std = np.std(residuals_i) + epsilon
+                    residuals_j_std = np.std(residuals_j) + epsilon
+                    Xs_i = X_i / (residuals_i_std + epsilon)
+                    Xs_j = X_j / (residuals_j_std + epsilon)
+                    Xs_im_mean = np.mean(Xs_i)
+                    Xs_jm_mean = np.mean(Xs_j)
+                    # Sum off-diagonal variance of empirical correlation
+                    w = (Xs_i - Xs_im_mean) * (Xs_j - Xs_jm_mean)
+                    w_mean = np.mean(w)
+                    sum_var_emp_corr += factor_var_emp_cor * np.sum(np.square(w - w_mean))
+                    # Sum squared empirical correlation
+                    sum_sq_emp_corr += np.square(factor_emp_cov * n_samples * w_mean)
+
+    # Calculate shrinkage intensity 
+    shrinkage = 1.0 - max(min((sum_var_emp_corr) / (sum_sq_emp_corr + epsilon), 1.0), 0.0)
+
+    # Shrink the empirical covariance
+    for i in prange(n_timeseries):
+        for j in range(i + 1):
+            if i != j:    
+                W[i, j] = W[j, i] = shrinkage * W[i, j]
+            else:
+                W[i, j] = W[j, i] = max(W[i, j], mint_shr_ridge)
+
+    return W
+
+# %% ../nbs/src/utils.ipynb 47
+# Lasso cyclic coordinate descent
+@njit("Array(float64, 1, 'C')(Array(float64, 2, 'C'), Array(float64, 1, 'C'), float64, int64, float64)", nogil=NUMBA_NOGIL, cache=NUMBA_CACHE, fastmath=NUMBA_FASTMATH, error_model="numpy")
+def _lasso(X: np.ndarray, y: np.ndarray, 
```

Comment:
Is this function working as expected? I get either very big numbers of zeros.

Reply:
This is a very valuable comment, thanks - this led to a rabbit hole uncovering that ERM-reg basically doesn't work at all, and it's also never tested (and was never tested before) - our unit test only tests ERM-reg_bu which does work. I looked back into old PRs and I don't think ERM-reg has ever worked.

To answer your questions:
- Zeros can be expected due to nan values present in the input data that ERM can't handle (I can think of how to handle these, but since ERM is essentially imho a fairly useless method as it's so computationally expensive I'm ok with this behaviour).
- If you'd perform a test without nan values you should see a P that's not full of zeros. 

I'll add a test for this and see if I can fix ERM-reg

Reply:
Btw these things are the reason I added much more methods to the model_performance test, as our testing is very limited. But I think I'll add even more tests, seeing these kinds of bugs.

Reply:
ERM-reg is basically garbage, I think we should leave it like this for now (old & new give same results)